### PR TITLE
Merge PR Branch to Migration Branch

### PR DESCRIPTION
**Pull request by noti0na1**
The new flag and NotNull list are rewritten in new style.

I made one change to the behavior of the NotNull list. I will comment it below.

We should merge #35 first and update this branch.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala [hunk 1]
```diff
@@ -57,117 +57,26 @@ object JavaNullInterop {
...
+    if (sym.name == nme.TYPE_ || sym.isAllOf(Flags.JavaEnumValue))
+      // Don't nullify the `TYPE` field in every class and Java enum instances
+      tp
+    else if (sym.name == nme.toString_ || sym.isConstructor || hasNotNullAnnot(sym))
+      // Don't nullify the return type of the `toString` method.
+      // Don't nullify the return type of constructors.
+      // Don't nullify the return type of methods with a not-null annotation.
+      nullifyExceptReturnType(tp)
+    else if (ctx.settings.YJavaInteropDontNullifyOutermost.value) 
+      nullifyExceptReturnType(tp)
```
Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala]: I think we can combine the method case and the field case now.

File: compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala [hunk 2]
```diff
@@ -260,38 +168,50 @@ object JavaNullInterop {
...
+        case None => false
       }
-    } else {
+    else 
       stats.getField(sym, tp) match {
-        case Some(fstas) =>
-          NoOpPolicy(_ == sym)
-        case None => FalsePolicy
-      }
+        case Some(_) => true
```
Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala:180]: For fields in the list, we should still nullify the inside type without a `| JavaNull` at the outermost, similar to `@NotNulll` annotation on a field.

**Comment by abeln** [comment]
#35 is now merged

File: compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala (hunk 1 above)
Comment (abeln) [reply compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala]: Right, I agree.

File: compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala [hunk 3]
```diff
@@ -57,117 +57,26 @@ object JavaNullInterop {
...
+      // Don't nullify the `TYPE` field in every class and Java enum instances
+      tp
+    else if (sym.name == nme.toString_ || sym.isConstructor || hasNotNullAnnot(sym))
+      // Don't nullify the return type of the `toString` method.
+      // Don't nullify the return type of constructors.
+      // Don't nullify the return type of methods with a not-null annotation.
+      nullifyExceptReturnType(tp)
+    else if (ctx.settings.YJavaInteropDontNullifyOutermost.value) 
+      nullifyExceptReturnType(tp)
+    else if (isInNotNullStdLibList(sym, tp))
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala]: We can merge this branch with the previous one.

File: compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala (hunk 2 above)
Comment (abeln) [reply compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala:180]: Yes, you're right. The previous code was incorrect. Nice catch!

**Review by abeln — APPROVED** [review]
This looks good. Before merging, could you please test it by re-running the community build and checking that the errors counts match what we had before?

**Comment by noti0na1** [comment]
@abeln I select several projects to test (scalacheck, scalatest, betterfile). The error counts are all different. 
Error counts (checker framework metadata) are slightly higher (except scalatest which has a lot more errors now 46 -> 226). 
The Error counts (java-interop-dont-nullify-outermost) are slightly lower.
I also tested the new project: intent, it has 10 and 5 errors without and with the flag respectively.

**Comment by abeln** [comment]
@noti0na1 thanks for checking

Could you please do the following
1) Pick any checker framework project (preferably one with fewer errors)
2) Run it with the branch before this PR and record the errors
3) Run it with the branch after this PR and record the errors
4) As per your message, there should be more errors in 3) than in 2)

What we want to understand is whether there are more errors because a problem in the merge, or because a problem in the _previous_ branch (which we already suspect to be the case because of the error you fix in JavaNullInterop).

If you could look into that, that would be great. If you could send me the error logs for 2) and 3), I can take a look as well.

Thanks!

cc @olhotak

**Comment by noti0na1** [comment]
@abeln I just found another potential bug in `JavaNullMap` of `JavaNullInterop`.

```scala
def needsNullArgs(tp: AppliedType): Boolean = !tp.classSymbol.is(JavaDefined)

// map
case appTp @ AppliedType(tycons, targs) =>
  val oldOutermostNullable = outermostLevelAlreadyNullable
  outermostLevelAlreadyNullable = false
  val targs2 = if (needsNullArgs(appTp)) targs map this else targs
  outermostLevelAlreadyNullable = oldOutermostNullable
  val appTp2 = derivedAppliedType(appTp, tycons, targs2)
  if (needsNull(tycons)) toJavaNullableUnion(appTp2) else appTp2
```

When nullify `List[Array[String]]`, the result would be `List[Array[String]]`, since `List` is Java Defined and `needsNullArgs(appTp) == false`. But I think a correct answer should be `List[Array[String | JavaNull]]`.

A fix:

```scala
case appTp @ AppliedType(tycons, targs) =>
  val oldOutermostNullable = outermostLevelAlreadyNullable
  // change this two lines
  outermostLevelAlreadyNullable = tp.classSymbol.is(JavaDefined)
  val targs2 = targs map this
  outermostLevelAlreadyNullable = oldOutermostNullable
  val appTp2 = derivedAppliedType(appTp, tycons, targs2)
  if (needsNull(tycons)) toJavaNullableUnion(appTp2) else appTp2
```

**Comment by abeln** [comment]
@noti0na1 good find!
Yes, I agree that we have a bug, and I agree with your fix. Could you please submit that fix as a separate PR to the explicit nulls branch. Could you please add your test as well as a few more
that exercise increased nestedness levels?

e.g.

```Scala
List[Array[List[Array[String]]]] => List[Array[List[Array[String|JavaNull]]|JavaNull]]|JavaNull
```

Thanks!

**Comment by abeln** [comment]
This will need to be updated after https://github.com/abeln/dotty/pull/37 is merged, right?

**Comment by noti0na1** [comment]
@abeln updated

**Comment by abeln** [comment]
Did you do a full test run for this? Do all tests pass?

**Comment by noti0na1** [comment]
@abeln Yes, all tests pass

**Comment by abeln** [comment]
Great! Thanks for doing this. When you have a chance, could you please do a full run of the community build and update the spreadsheet? We're mainly interested in the errors counts _with_ the don't-nullify-java-types flag enabled.

**Pull request merged**